### ISSUE TITLE
Failure to load using explicit file: URI scheme

### ISSUE THREAD

**Issue by gotthardp** [openssl/openssl#15922]
May I use the `file:` URI scheme explicitly? I believe it used to work some months ago, but it doesn't right now.

```
$ openssl genrsa -out pubkey.pem
$ openssl pkey -in file:pubkey.pem
C077A7B6647F0000:error:80000002:system library:file_open:No such file or directory::0:calling stat(file:pubkey.pem)
C077A7B6647F0000:error:1C8000DB:Provider routines:file_open:path must be absolute::0:Given path=pubkey.pem
```

**Comment by levitte** [comment]
The [`file:` scheme](https://datatracker.ietf.org/doc/html/rfc8089#section-2) doesn't allow relative paths, and our loaders have been coded accordingly.  You might think of `OSSL_STORE_attach()`, which allows you to pass a BIO and a scheme to associate it with.  Do note that I don't expect other store loaders than for the `file:` scheme to support this feature...

So, the proper way to do what you want is `openssl pkey -in file:$(pwd)/pubkey.pem`

**Comment by gotthardp** [comment]
But `openssl pkey -in pubkey.pem` works even with the relative path. That's also correct?

**Comment by levitte** [comment]
> But `openssl pkey -in pubkey.pem` works even with the relative path. That's also correct?

Yes.  For backward compatibility, we allow relative paths when there's no explicit scheme given.  It's the explicit `file:` that makes all the difference in interpretation.

**Comment by levitte** [comment]
> (But for users unaware of this it's quite surprising though.)

So this isn't sufficient (from the new `doc/man1/openssl-pkey.pod`)?

> **Input options**
> **-in** _filename_ | _uri_
> This specifies the input to read a key from or standard input if
> this option is not specified.  ...

**Comment by mattcaswell** [comment]
Do we have somewhere that describes what URI schemes we support and what the format of those URI schemes is?

**Comment by t-j-h** [comment]
> The [`file:` scheme](https://datatracker.ietf.org/doc/html/rfc8089#section-2) doesn't allow relative paths

Note that the original file scheme was explicitly **not** hardwired to absolute paths https://datatracker.ietf.org/doc/html/rfc1738#section-3.10 and that is what you will find most users remain conceptually used to. 

Nothing was actually said about absolute or relative with respect to the path in the RFC - it was just a "path".

**Comment by levitte** [comment]
> Do we have somewhere that describes what URI schemes we support and what the format of those URI schemes is?

We have `doc/man7/ossl_store-file.pod`...  which needs a bit of an update.

**Comment by levitte** [comment]
@t-j-h, what a relative path means is up for interpretation, or well, can be made relative to a base.

So for example, I just noticed that current Firefox does support relative paths, and has `/` as the base...  `file:foo.pem` gets rewritten to `file:///file.pem`, `file:foo/bar.pem` to `file:///foo/bar.pem`, `file:../foo/bar.pem` to `file:///foo/bar.pem`.
Is this what "most users remain conceptually used to"?
